### PR TITLE
Change mobile search filters modal to a regular 'in page' component

### PR DESCRIPTION
**Pull request by AshGDS**
## What

- Changes the Filters on mobile from being a modal/popup, to a regular "In Page" element thats visibility is triggered by a `<button>` styled as a link.
- Fixes a bug where the amount of filters you selected would not disappear after clicking "Clear all filters". This was because `<span>` was not included in elements to remove in the JS code for this.
- Fixes a visual bug where on Desktop/Tablet, the Topic and Subtopics filters had a `min-width` which would make them spill over onto the search results.

## Why
- The filters change was needed as part of our accessibility fixes.

## Testing

To test, you don't need any local static. You can just use `../startup.sh --live`. Then navigate to http://127.0.0.1:3062/search/all and reduce your screen size until it's a mobile view.

Accessibility wise, I have tested the keyboard functionality, MacOS VoiceOver, Android Talkback, JAWS, NVDA, Lighthouse, WebAim WAVE, navigating at 400% zoom, the behaviour when JavaScript is disabled, and checked the heading structure with headingsMap.

### Before - Mobile Filters
https://user-images.githubusercontent.com/8880610/219028588-90215c3a-a271-44fe-82dd-3827611d7d92.mov

### After - Mobile Filters

https://user-images.githubusercontent.com/8880610/219028825-f43558c7-69c5-4f81-8e06-8e8444823432.mov

## Before - Visual Bug
image

## After - Visual Bug

image

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

This is a draft PR for an accessibility fix, the code will change quite a bit and we cleaned up before it is ready for a PR.

### CODE REVIEW

**Comment by AshGDS** [comment]
Hi @andysellick - would you be able to review this? It's a bit of a big change so may be easier to look at it commit by commit or to look at `mobile-filters-modal.js` in your editor. It's my first time contributing SASS/HTML so hopefully everything is OK code wise. Thanks :+1:

File: app/assets/javascripts/modules/mobile-filters-modal.js
Comment (andysellick) [inline app/assets/javascripts/modules/mobile-filters-modal.js:28]: Not sure this check needs to be in place? Is there a situation where `event` wouldn't be defined?

File: app/assets/javascripts/modules/mobile-filters-modal.js
Comment (andysellick) [inline app/assets/javascripts/modules/mobile-filters-modal.js:33]: Why do you need to re-bind `handleClose` - can these lines not be `this.handleClose(event)` ?

File: app/assets/stylesheets/finder_frontend.scss
Comment (andysellick) [inline app/assets/stylesheets/finder_frontend.scss:359]: We shouldn't be styling components in applications - if this is a change that needs to be made to a component, it should be made in the gem, otherwise we're breaking [component isolation](https://github.com/alphagov/govuk_publishing_components/blob/main/docs/component_principles.md#a-component-is-isolated-when).

File: app/assets/stylesheets/finder_frontend.scss
Comment (andysellick) [inline app/assets/stylesheets/finder_frontend.scss:360]: A select component with the full_width option (which this is) already has width set to 100%, so there's something odd going on here. It looks like the select element itself is escaping from the bounds of the component itself, which I think should be investigated at the component level.

File: app/views/finders/_facet_collection.html.erb
Comment (andysellick) [inline app/views/finders/_facet_collection.html.erb:4]: Nitpick: indentation.

File: app/views/finders/_filter_button.html.erb
Comment (andysellick) [inline app/views/finders/_filter_button.html.erb:1]: This should be a class rather than an inline style - it's better practice and will soon be outlawed under our CSP.

File: app/assets/javascripts/modules/mobile-filters-modal.js
Comment (andysellick) [inline app/assets/javascripts/modules/mobile-filters-modal.js:11]: Not sure where to put this comment, so adding here to allow you to reply. I wanted to ask about the behaviour of the `Show search results` button. It seems to be a jump link to the search results immediately below it, which makes sense now I've thought about it but initially confused me as I assumed it would hide the search filters. Is this intended behaviour?

File: app/assets/javascripts/modules/mobile-filters-modal.js
Comment (andysellick) [inline app/assets/javascripts/modules/mobile-filters-modal.js:79]: Again not sure where to put this comment. On Chrome on the mobile view the contents of the select component appear very small and sometimes not apparently attached to the element itself. Not seeing the same thing in Firefox, so not sure what's going on, probably more to do with the component than this change. Something to investigate separately, perhaps.

Screenshot 2023-02-21 at 15 39 35

File: app/assets/javascripts/modules/mobile-filters-modal.js
Comment (AshGDS) [reply app/assets/javascripts/modules/mobile-filters-modal.js:11]: @andysellick Yep this is intentional - at first I removed the button entirely, and even though the results update automatically the designer thought it's better to have a button which jumps to the results.

File: app/assets/javascripts/modules/mobile-filters-modal.js
Comment (AshGDS) [reply app/assets/javascripts/modules/mobile-filters-modal.js:28]: @andysellick I believe it's because the tests call the functions without firing an event, so you wouldn't be able to call `.preventDefault()` on something that's undefined. Is this OK?

File: app/assets/javascripts/modules/mobile-filters-modal.js
Comment (AshGDS) [reply app/assets/javascripts/modules/mobile-filters-modal.js:33]: Good spot, I think it can be `this.module.close()` which is already binded to `this`. I think the binding is needed because within that close function, it calls `this.module` etc, and without the binding, the scope of `this` would be the function itself.

File: app/assets/stylesheets/finder_frontend.scss
Comment (AshGDS) [reply app/assets/stylesheets/finder_frontend.scss:360]: @andysellick Yep it has `width: 100%` on it, but `.govuk-select` has `min-width: 11.5em;` which causes the issue. Should I add the `min-width: 100%` to the select component instead?

File: app/views/finders/_filter_button.html.erb
Comment (AshGDS) [reply app/views/finders/_filter_button.html.erb:1]: Whoops, good spot, was meant to be temporary for testing!

File: app/assets/javascripts/modules/mobile-filters-modal.js
Comment (AshGDS) [reply app/assets/javascripts/modules/mobile-filters-modal.js:79]: @andysellick Are you using an external monitor? Mine looks like this 
image

**Comment by AshGDS** [comment]
Thanks @andysellick - I've resolved the changes I could make, but I have a question regarding the `min-width` thing. Is it easier to fix that as a separate PR instead?

**Comment by andysellick** [comment]
> Thanks @andysellick - I've resolved the changes I could make, but I have a question regarding the `min-width` thing. Is it easier to fix that as a separate PR instead?

Thanks, I'll do another review. Yes, probably best to sort out that issue in a separate PR, as I guess it might involve changing the component.

File: app/assets/stylesheets/finder_frontend.scss
Comment (andysellick) [reply app/assets/stylesheets/finder_frontend.scss:362]: Agree that it's probably easier to address this in a separate PR.

File: app/assets/javascripts/modules/mobile-filters-modal.js
Comment (andysellick) [reply app/assets/javascripts/modules/mobile-filters-modal.js:11]: Thanks. I think now that this might be a significant enough change that we should do some user testing to make sure we've got it right. That would make this piece of work a lot bigger and more complex though. I'll put a note on the trello card and we can discuss it further offline.

File: app/assets/javascripts/modules/mobile-filters-modal.js
Comment (andysellick) [reply app/assets/javascripts/modules/mobile-filters-modal.js:68]: It happens when I switch to mobile view in Chrome but keep the window wide enough to still see the desktop site. Have raised an issue about it to avoid blocking this PR: https://github.com/alphagov/govuk_publishing_components/issues/3280

File: app/views/finders/_facet_collection.html.erb
Comment (andysellick) [inline app/views/finders/_facet_collection.html.erb:8]: Nitpick: there still seems to be something wrong with the indentation here.

File: app/views/finders/_facet_collection.html.erb
Comment (andysellick) [inline app/views/finders/_facet_collection.html.erb:2]: Is this `if` block over-complicated? Could it be written instead as

```Ruby
<% if content_item.all_content_finder? or facets.select(&:filterable?).any? %>
  <%= render partial: 'filter_button'%>
<% end %>
```

File: app/views/finders/_facet_collection.html.erb
Comment (AshGDS) [reply app/views/finders/_facet_collection.html.erb:8]: Whoops, my bad!

File: app/views/finders/_facet_collection.html.erb
Comment (AshGDS) [reply app/views/finders/_facet_collection.html.erb:2]: No worries, yeah I didn't change any Ruby logic in this PR luckily! If it was on one line, wouldn't it need to be:

```Ruby
<% if content_item.all_content_finder? or !content_item.all_content_finder? and facets.select(&:filterable?).any? %>
  <%= render partial: 'filter_button'%>
<% end %>
```

as in the original, the `elsif` only runs if `content_item.all_content_finder?` is `false`, whereas that simplified solution wouldn't check for that.

File: app/assets/stylesheets/finder_frontend.scss
Comment (AshGDS) [reply app/assets/stylesheets/finder_frontend.scss]: Thanks, I removed it for now.

File: app/assets/javascripts/modules/mobile-filters-modal.js
Comment (AshGDS) [reply app/assets/javascripts/modules/mobile-filters-modal.js:11]: @andysellick  The UI changes were posted in the accessibility community, and there were no major issues found - initially there were some changes requested, but they were rescinded after further investigation. Therefore, it has had some internal testing already. Before it was rescinded, one of the suggestions was to change the text to something like "Go to search results" - maybe we changing the text would be useful? Or "Jump to search results" or "Skip to search results"?

**Comment by AshGDS** [comment]
Thanks @andysellick - I removed the UI bug fix, fixed the indentation, and responded to the user testing suggestion.

File: app/assets/javascripts/modules/mobile-filters-modal.js
Comment (andysellick) [reply app/assets/javascripts/modules/mobile-filters-modal.js:11]: `Go to search results` sounds good, let's change it to that.

File: app/views/finders/_facet_collection.html.erb
Comment (andysellick) [reply app/views/finders/_facet_collection.html.erb:8]: Ummm... still not quite right. The inner lines are partly indented by 2 spaces, partly by 4. Can you take another look?

Comment (AshGDS) [reply app/views/finders/_facet_collection.html.erb:8]: @andysellick My bad, I've been using the `tab` key on Visual Studio Code to indent this, so I'm guessing it's judgement is wrong with `.erb` files.

File: app/views/finders/_facet_collection.html.erb
Comment (andysellick) [reply app/views/finders/_facet_collection.html.erb:2]: Yes, I think my solution was a little too simple, but isn't the elsif unnecessary? The code is basically saying

```
if a_thing
  (render partial)
elsif not a_thing
  if another_thing
    (render partial)
  end
end
```

i.e. there are only two possible outcomes for the top level if statement - a_thing is either or not, so it's always going to run one of those statements. Doesn't that mean it could be

```
if a_thing
  (render partial)
elsif another_thing
  (render partial)
end
```

I concede that I still might be missing something, it's Friday afternoon, so happy for you to point out the flaw in my logic.

File: app/views/finders/_facet_collection.html.erb
Comment (AshGDS) [reply app/views/finders/_facet_collection.html.erb:8]: Actually it's probably my fault. @andysellick Does it look OK now? If I got it wrong again, may be worth writing out the correct indentation 😅

File: app/views/finders/_facet_collection.html.erb
Comment (AshGDS) [reply app/views/finders/_facet_collection.html.erb:2]: @andysellick That example makes sense - I think because your original suggestion uses `or` it would've changed the flow to:

```
    if a_thing or another_thing
        (render partial)
```

instead of 

```
    if a_thing
      (render partial)
    elsif another_thing
      (render partial)
    end
```

and no worries, I have Friday brain too

Comment (andysellick) [reply app/views/finders/_facet_collection.html.erb:2]: Thanks for discussing this with me just now Ash. As discussed, I think we both understand this a bit more clearly but on the basis that this is a bit of code that's been moved rather than written as part of this PR, I think we should leave it as it is for now.

**Review by andysellick — APPROVED** [review]
All good, happy to approve. One thing before you merge, can any of the commits be squashed together?

**Comment by AshGDS** [comment]
Thanks @andysellick - will do :+1:

**Pull request merged**
